File: backend/fastapi_app/schemas/progress.py
"""
Progress schemas matching Django serializers.
"""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class ProgressBase(BaseModel):
    """
    Base progress schema matching Django ProgressBaseSchema.
    """
    content_id: str
    content_type: str
    status: Literal["not_started", "in_progress", "completed", "paused"]
    completion_percentage: int = 0
    score: float | None = None
    metadata: dict = {}


class ProgressResponse(ProgressBase):
    """
    Progress response schema matching Django ProgressResponseSchema.
    """
    id: int
    user_id: int
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True,
    }

File: frontend/nextjs_app/services/types/progress.ts
/**
 * Progress types matching Django DRF ProgressSerializer
 * These types should stay in sync with backend/django_app/progress/serializers.py
 */

export interface Progress {
  id: number;
  user: number;
  content_id: string;
  content_type: string;
  status: 'not_started' | 'in_progress' | 'completed' | 'paused';
  completion_percentage: number;
  score?: number;
  started_at?: string;
  completed_at?: string;
  metadata: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface CreateProgressRequest {
  content_id: string;
  content_type: string;
  status?: 'not_started' | 'in_progress' | 'completed' | 'paused';
  completion_percentage?: number;
  metadata?: Record<string, any>;
}

export interface UpdateProgressRequest {
  status?: 'not_started' | 'in_progress' | 'completed' | 'paused';
  completion_percentage?: number;
  score?: number;
  metadata?: Record<string, any>;
}

